File: src/main/java/virtualMachine/vm_instruction_processing/impl/UnaryInstructionProcessor.java
package virtualMachine.vm_instruction_processing.impl;

import virtualMachine.stack.memory.VirtualMemory;
import virtualMachine.stack.types.instruction.Commands;
import virtualMachine.stack.types.instruction.Instruction;
import virtualMachine.vm_instruction_processing.InstructionProcessor;

public class UnaryInstructionProcessor implements InstructionProcessor {

    @Override
    public void processInstruction(Instruction instruction, VirtualMemory virtualMemory) {
        var head = virtualMemory.pop();

        if (instruction.getCommand() == Commands.NOT) {
            virtualMemory.push(head.bitWiseNot());
        } else if (instruction.getCommand() == Commands.NEGATIVE) {
            virtualMemory.push(head.negate());
        }
    }
}

File: src/main/kotlin/virtualMachine/stack/vm_instruction_parsing/vm_instruction_processing/impl/UnaryInstructionProcessor.kt
package virtualMachine.stack.vm_instruction_parsing.vm_instruction_processing.impl

import virtualMachine.stack.memory.GlobalVirtualMemory
import virtualMachine.stack.datawrappers.SixteenBit
import virtualMachine.stack.vm_instruction_parsing.vm_instruction_processing.InstructionProcessor

class UnaryInstructionProcessor : InstructionProcessor {

    override fun processInstruction(instruction: String, virtualMemory: GlobalVirtualMemory) {
        val uniaryInstruction: SixteenBit = virtualMemory.popLocalStack()

        if (instruction == "not") {
            virtualMemory.pushToLocalStack(uniaryInstruction.not16Bit())
        }
    }
}